### PR TITLE
Add support for adding private binaries into scope

### PR DESCRIPTION
**Pull request by rgrinberg**
This PR adds support for overriding the binary in a scoped way. For example:

```
(env
 (_
  (bins (./foo/bar.exe as bar.exe))))
```

### TODO:

- [x] I just noticed that cross compilation is a bit broken currently. For example, `resolve_program` does not seem to respect it. I will fix this issue (and others like it).

- [x] `PATH` is not being modified. We've agreed that w/e binaries were added into scope using `(bins ..)` should also be added to `PATH` (without any sandboxing). This is not so simple to implement, so I might delay this to another PR.

Copying users who have shown interest in this feature: @ejgallego @samoht and @avsm

### CODE REVIEW

**Comment by diml** [comment]
Regarding `PATH`, it is indeed the case that it shouldn't be modified for individual rules, however a new directory should be added everytime we see a `(bins ...)`. Otherwise, we won't be able to use he same name in different directories, for instance:

```
foo/dune:
(env (_ (bins (foo.exe as x)))

bar/dune:
(env (_ (bins (bar.exe as x)))
```

BTW, in addition to this feature, I'd like that we also support declaring that a binary is available in the current project via the `executable` stanza directly. The reason is that in general it is not necessary in Dune to specify explicit paths, while this is not true for this feature. As a result, in your example if you rename `foo` to `foo2`, then you also have to edit the path in the `env` stanza, which is a but unfortunate.

**Comment by ejgallego** [comment]
I guess this only works for users of `(run foo)` right?

My use case was a bit different, as I wanted to add a binary to an action, in this case the action building the sphinx docs so it could locate where coqtop is.

**Comment by rgrinberg** [comment]
Well, this is the first step. Once I finish the 2nd todo point, the binary will in fact be in `PATH`. All that would be left would be to add support for `%{bin:..}` in deps.

**Comment by diml** [comment]
BTW, this is touching the same code as #1354. Looking at the two features, I believe it would be easier to do the changes discussed in #1354 for the artifacts.ml module first. The general idea is that instead of manually building the list of local binaries by scanning the stanzas, we instead list the contents of the local bin directory.

**Comment by rgrinberg** [comment]
This PR is now rebased on master (which includes #1354). In the process, I've also allowed for some variable expansion in this `bins` field.

> The general idea is that instead of manually building the list of local binaries by scanning the stanzas, we instead list the contents of the local bin directory.

Do you think we should do this for the "local" (dir/env specific) binary directory as well? It seems like there's not as much benefit here as `File_bindings.t` is a one to one mapping to the contents of this directory.

**Comment by rgrinberg** [comment]
Okay, now everything works. The code is a bit of a mess right now, but it's ready for another round of review.

The last commit makes passing PATH a lot more explicit at the cost of boilerplate. In my opinion, it's worth it because it was too easy before to look up a binary without accounting for the env overrides. Now we should be able to override almost any binary in a build in a scoped way.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/print_diff.ml [hunk 1]
```diff
@@ -21,7 +21,8 @@ let print ?(skip_trailing_cr=Sys.win32) path1 path2 =
       (Path.to_string_maybe_quoted path2)
   in
   let normal_diff () =
-    match Bin.which "diff" with
+    (* TODO we should be using the diff program from env *)
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/print_diff.ml]: Yeah, which is why I left it as a TODO.

File: src/env_node.ml [hunk 2]
```diff
@@ -18,8 +20,25 @@ let make ~dir ~inherit_from ~scope ~config ~env =
...
+  ; artifacts = None
+  ; file_bindings = None
   }
 
+let file_bindings t ~profile ~expander =
+  match t.file_bindings with
+  | Some x -> x
+  | None ->
+    let file_bindings =
+      match Dune_env.Stanza.find t.config ~profile with
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/env_node.ml]: Yeah, that's a good point. Will do it in a subsequent PR.

File: src/print_diff.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (diml) [reply src/print_diff.ml]: I mean, maybe it shouldn't even be a TODO. It adds more error cases to handle: for instance if the construction of the `diff` tool fails, then you end up in an awkward state. Not that it's not feasible to handle it, but I'm not sure it's worth making the code more complex than it already is.

File: src/artifacts.ml [hunk 3]
```diff
@@ -49,6 +49,18 @@ let binary t ?hint ~loc name =
...
 
+let add_binaries t ~dir = function
+  | [] -> t
+  | bindings ->
+    let local_bins =
+      lazy (
+        List.fold_left bindings ~init:(Lazy.force t.local_bins)
+          ~f:(fun acc fb ->
+            let path = File_bindings.dst_path fb ~dir:(Utils.local_bin dir) in
+            String.Map.add acc (Path.basename path) path))
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/artifacts.ml]: I guess. But I'm a little confused about how we want this to work. We want the keys in this map to always be without the `.exe` extension, right? Then why do we only want to strip it on Windows?

I've added the .exe stripping when the `as` name isn't specified. So now the syntax is:

```
(foo.exe as bar)
```

But now I find that it's awkward that we have a target on the left and a path on the right. Do you think it's worth stripping an exe from the `as <name>` as well to have:

```
(foo.exe as bar.exe)
```

**Comment by rgrinberg** [comment]
@diml this is ready for another round of review

**Review by ghost — APPROVED** [review]
Looks good

**Comment by rgrinberg** [comment]
It's updated.

**Pull request merged**